File: python-fastapi/app/routes/coverage.py
import json
from datetime import date
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, Request, Response
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.fhir_mapper import coverage_to_fhir, search_bundle
from app.models.coverage import CoverageModel
from app.models.patient import PatientModel
from app.services.coverage_service import save_coverage

router = APIRouter(prefix="/fhir/Coverage", tags=["Coverage"])
FHIR_JSON = "application/fhir+json"


@router.get("/{coverage_id}")
async def read_coverage(coverage_id: UUID, db: AsyncSession = Depends(get_db)):
    result = await db.execute(select(CoverageModel).where(CoverageModel.id == coverage_id))
    coverage = result.scalars().first()
    if not coverage:
        raise HTTPException(status_code=404, detail="Coverage not found")
    return Response(content=json.dumps(coverage_to_fhir(coverage), default=str), media_type=FHIR_JSON)


@router.get("")
async def search_coverages(
    beneficiary: UUID | None = None, db: AsyncSession = Depends(get_db)
):
    if beneficiary:
        result = await db.execute(
            select(CoverageModel).where(CoverageModel.patient_id == beneficiary)
        )
    else:
        result = await db.execute(select(CoverageModel))
    coverages = result.scalars().all()
    fhir_resources = [coverage_to_fhir(c) for c in coverages]
    return Response(
        content=json.dumps(search_bundle(fhir_resources), default=str),
        media_type=FHIR_JSON,
    )


@router.post("", status_code=201)
async def create_coverage(request: Request, db: AsyncSession = Depends(get_db)):
    body = await request.json()

    patient_ref = body["beneficiary"]["reference"]
    patient_id = UUID(patient_ref.replace("Patient/", ""))
    result = await db.execute(select(PatientModel).where(PatientModel.id == patient_id))
    if not result.scalars().first():
        raise HTTPException(status_code=400, detail=f"Patient not found: {patient_id}")

    payor = body["payor"][0]
    period = body.get("period", {})

    entity = CoverageModel(
        patient_id=patient_id,
        subscriber_id=body.get("subscriberId", ""),
        status=body.get("status", "active"),
        payor_name=payor.get("display", ""),
        payor_id=payor.get("reference", "").replace("Organization/", ""),
        plan_name=payor.get("display", ""),
        relationship=body.get("relationship", {}).get("coding", [{}])[0].get("code", "self"),
        period_start=date.fromisoformat(period["start"]) if "start" in period else None,
        period_end=date.fromisoformat(period["end"]) if "end" in period else None,
    )

    saved = await save_coverage(db, entity)
    return Response(
        content=json.dumps(coverage_to_fhir(saved), default=str),
        media_type=FHIR_JSON,
        headers={"Location": f"/fhir/Coverage/{saved.id}"},
    )

File: node-express/src/controllers/coverage.js
const { Router } = require('express');
const pool = require('../config/database');
const coverageService = require('../services/coverageService');
const { coverageToFhir, searchBundle } = require('../models/fhirMapper');

const router = Router();
const FHIR_JSON = 'application/fhir+json';

router.get('/:id', async (req, res) => {
  const coverage = await coverageService.findById(req.params.id);
  if (!coverage) return res.status(404).json({ error: 'Coverage not found' });
  res.type(FHIR_JSON).json(coverageToFhir(coverage));
});

router.get('/', async (req, res) => {
  let coverages = [];
  if (req.query.beneficiary) {
    coverages = await coverageService.findByPatientId(req.query.beneficiary);
  }
  res.type(FHIR_JSON).json(searchBundle(coverages.map(coverageToFhir)));
});

router.post('/', async (req, res) => {
  const body = req.body;
  const patientId = body.beneficiary.reference.replace('Patient/', '');

  const patientCheck = await pool.query('SELECT 1 FROM patients WHERE id = $1', [patientId]);
  if (!patientCheck.rows[0]) return res.status(400).json({ error: `Patient not found: ${patientId}` });

  const payor = body.payor[0];
  const period = body.period || {};

  const saved = await coverageService.saveCoverage({
    patient_id: patientId,
    subscriber_id: body.subscriberId || '',
    status: body.status || 'active',
    payor_name: payor.display || '',
    payor_id: (payor.reference || '').replace('Organization/', ''),
    plan_name: payor.display || '',
    group_number: null,
    relationship: body.relationship?.coding?.[0]?.code || 'self',
    period_start: period.start || null,
    period_end: period.end || null,
  });

  res.status(201).type(FHIR_JSON)
    .set('Location', `/fhir/Coverage/${saved.id}`)
    .json(coverageToFhir(saved));
});

module.exports = router;
